### PR TITLE
Track angle histogramming and plotting

### PR DESCRIPTION
**Pull request by laborleben**

### CODE REVIEW

File: testbeam_analysis/result_analysis.py [hunk 1]
```diff
@@ -848,3 +849,56 @@ def calculate_efficiency(input_tracks_file, input_alignment_file, output_pdf, bi
                         out_pass[:] = total_track_density_with_DUT_hit.T
                         out_total[:] = total_track_density.T
     return efficiencies, pass_tracks, total_tracks
+
+
+def track_angle(input_tracks_file, output_track_angle_file, use_duts=None, output_pdf=None):
+    '''Calculates and histograms the track angle of the fitted tracks for selected DUTs.
```
Comment (DavidLP) [inline testbeam_analysis/result_analysis.py]: Good that you added docstrings

File: testbeam_analysis/result_analysis.py [hunk 2]
```diff
@@ -848,3 +849,56 @@ def calculate_efficiency(input_tracks_file, input_alignment_file, output_pdf, bi
...
+                   ' ', progressbar.AdaptiveETA()]
+        progress_bar = progressbar.ProgressBar(widgets=widgets,
+                                               maxval=len(use_duts),
+                                               term_width=80)
+        progress_bar.start()
+
+        i = 0  # counting number of progressed DUTs, needed for progressbar
+        for node in in_file_h5.root:  # loop through all DUTs in track table
+            if node.name in dut_index:  # only store track slopes of selected DUTs
+                slopes.update({node.name: [node[::10]['slope_0'],
```
Comment (DavidLP) [inline testbeam_analysis/result_analysis.py]: This is a little hackish. You just load every 10th track into RAM. No guarantee that it fits and a lot of tracks a not used. Imagine you only have 100 tracks. Please take the complete table. This table should be read in chunks as it is done in the [other](https://github.com/SiLab-Bonn/testbeam_analysis/blob/development/testbeam_analysis/result_analysis.py#L709) result analysis function. There is a helper function `data_aligned_at_events` that you could use. Although event aligning is not really needed here.

File: testbeam_analysis/result_analysis.py [hunk 3]
```diff
@@ -848,3 +849,56 @@ def calculate_efficiency(input_tracks_file, input_alignment_file, output_pdf, bi
...
+                                           node[::10]['slope_1']]})
+                i = i + 1
+                progress_bar.update(i)
+        progress_bar.finish()
+
+    result = np.zeros(shape=(n_duts,), dtype=[('mean_slope_0', np.float), ('mean_slope_1', np.float), ('sigma_slope_0', np.float), ('sigma_slope_1', np.float)])
```
Comment (DavidLP) [inline testbeam_analysis/result_analysis.py]: I would call it x and y instead of 0, 1

File: testbeam_analysis/tools/plot_utils.py [hunk 4]
```diff
@@ -1055,3 +1055,37 @@ def efficiency_plots(hit_hist, track_density, track_density_with_DUT_hit, effici
...
+def plot_track_angle(input_track_slopes, output_data, output_pdf):
+    direction = ['x', 'y', 'z']
+
+    with PdfPages(output_pdf) as output_fig:
+        i = 0  # counting number of progressed DUTs
+        for key in input_track_slopes:
+            for j in (0, 1):  # iterate over slopes
+                min = np.min(input_track_slopes[key][j])
+                max = np.max(input_track_slopes[key][j])
+                edges = np.arange(min - 0.00005, max + 0.00005, 0.0001)
```
Comment (DavidLP) [inline testbeam_analysis/tools/plot_utils.py]: Please do not hard code. Instead define a `bin_size_x`, `bin_size_y` in the function

File: testbeam_analysis/tools/plot_utils.py [hunk 5]
```diff
@@ -1055,3 +1055,37 @@ def efficiency_plots(hit_hist, track_density, track_density_with_DUT_hit, effici
...
+                bin_center = (edges[1:] + edges[:-1]) / 2.0
+
+                sigma = np.std(np.arctan(input_track_slopes[key][j]))  # initial guess for sigma
+                mean = np.mean(np.arctan(input_track_slopes[key][j]))  # initial guess for mean
+
+                histo = plt.hist(np.arctan(input_track_slopes[key][j]), edges, label='Angular Distribution', color='#4361ba')
```
Comment (DavidLP) [inline testbeam_analysis/tools/plot_utils.py]: To unify the colors one could use a 'c' color here. Check [matplotlib 2 color nfo](http://matplotlib.org/api/colors_api.html).

File: testbeam_analysis/tools/plot_utils.py [hunk 6]
```diff
@@ -1055,3 +1055,37 @@ def efficiency_plots(hit_hist, track_density, track_density_with_DUT_hit, effici
         output_fig.savefig()
     else:
         logging.warning('Cannot create efficiency plots, since all pixels are masked')
+
+
+def plot_track_angle(input_track_slopes, output_data, output_pdf):
+    direction = ['x', 'y', 'z']
+
+    with PdfPages(output_pdf) as output_fig:
+        i = 0  # counting number of progressed DUTs
```
Comment (DavidLP) [inline testbeam_analysis/tools/plot_utils.py]: In python one rarely does use a stand alone index variable. More common is the 
```
for value in iterable: 
  do something
```
pattern

File: testbeam_analysis/result_analysis.py [hunk 7]
```diff
@@ -847,3 +848,55 @@ def calculate_efficiency(input_tracks_file, input_alignment_file, output_pdf, bi
...
+        else:
+            use_duts = use_duts
+        dut_index = [('Tracks_DUT_%i' % i) for i in use_duts]
+
+        slope_0_temp = np.array([])
+        slope_1_temp = np.array([])
+        for node in in_file_h5.root:  # loop through all DUTs in track table
+            if node.name in dut_index:
+                for tracks_chunk, _ in analysis_utils.data_aligned_at_events(node, chunk_size=100000):  # only store track slopes of selected DUTs
+                    slope_0_chunk = tracks_chunk[:]['slope_0']
```
Comment (DavidLP) [inline testbeam_analysis/result_analysis.py]: `[:]` is not needed here

File: testbeam_analysis/result_analysis.py [hunk 8]
```diff
@@ -847,3 +848,55 @@ def calculate_efficiency(input_tracks_file, input_alignment_file, output_pdf, bi
...
+            use_duts = range(n_duts)
+        else:
+            use_duts = use_duts
+        dut_index = [('Tracks_DUT_%i' % i) for i in use_duts]
+
+        slope_0_temp = np.array([])
+        slope_1_temp = np.array([])
+        for node in in_file_h5.root:  # loop through all DUTs in track table
+            if node.name in dut_index:
+                for tracks_chunk, _ in analysis_utils.data_aligned_at_events(node, chunk_size=100000):  # only store track slopes of selected DUTs
```
Comment (DavidLP) [inline testbeam_analysis/result_analysis.py]: Yes much better. Chunk size should be a parameter of the function as it is for all other result analysis functions. A chunk_size of 499999 is usually used

File: testbeam_analysis/tools/plot_utils.py [hunk 9]
```diff
@@ -1067,3 +1067,36 @@ def efficiency_plots(hit_hist, track_density, track_density_with_DUT_hit, effici
...
         logging.warning('Cannot create efficiency plots, since all pixels are masked')
+
+
+def plot_track_angle(input_track_slopes, output_data, output_pdf, use_n_duts, n_bins):
+    direction = ['x', 'y', 'z']
+
+    with PdfPages(output_pdf) as output_fig:
+        for i in range(use_n_duts):  # counting number of progressed DUTs
+            for key in input_track_slopes:
+                for j in (0, 1):  # iterate over slopes
```
Comment (DavidLP) [inline testbeam_analysis/tools/plot_utils.py]: iterate over slopes _direction_ ? No?

Comment (YannickDieter) [inline testbeam_analysis/tools/plot_utils.py]: Sorry what do you mean with this? j is the index for iterating over the slope directions.

File: testbeam_analysis/result_analysis.py (hunk 8 above)
Comment (laborleben) [inline testbeam_analysis/result_analysis.py]: ...and should be a parameter in the histogram_track_angle(..., chunk_size=499999).

File: testbeam_analysis/tools/plot_utils.py (hunk 9 above)
Comment (DavidLP) [inline testbeam_analysis/tools/plot_utils.py]: Your comment says you iterate over the slopes, but you iterate over the slope _directions_

File: testbeam_analysis/result_analysis.py (hunk 8 above)
Comment (DavidLP) [inline testbeam_analysis/result_analysis.py]: As I said ;-)

**Comment by DavidLP** [comment]
Looks good, fine to merge for me. Would be nice to have this analysis in an example. @laborleben You merge?

**Comment by YannickDieter** [comment]
Perfect. Maybe I could push my analysis script as an example.

**Comment by DavidLP** [comment]
If it justs adds the purity calculation without changing the other stuff go for it

**Pull request merged**